File: cpp/model/surfsys.cpp
////////////////////////////////////////////////////////////////////////////////
// STEPS - STochastic Engine for Pathway Simulation
// Copyright (C) 2007-2009ÊOkinawa Institute of Science and Technology, Japan.
// Copyright (C) 2003-2006ÊUniversity of Antwerp, Belgium.
//
// See the file AUTHORS for details.
//
// This file is part of STEPS.
//
// STEPSÊis free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// STEPSÊis distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.ÊIf not, see <http://www.gnu.org/licenses/>.
//
////////////////////////////////////////////////////////////////////////////////

/*
 *  Last Changed Rev:  $Rev: 307 $
 *  Last Changed Date: $Date: 2010-03-24 09:25:37 +0900 (Wed, 24 Mar 2010) $
 *  Last Changed By:   $Author: wchen $
 */


// STL headers.
#include <cassert>
#include <sstream>
#include <string>
#include <map>

// STEPS headers.
#include "../common.h"
#include "../error.hpp"
#include "model.hpp"
#include "spec.hpp"
#include "surfsys.hpp"
#include "sreac.hpp"

////////////////////////////////////////////////////////////////////////////////

USING_NAMESPACE(std);
USING_NAMESPACE(steps::model);

////////////////////////////////////////////////////////////////////////////////

Surfsys::Surfsys(string const & id, Model * model)
: pID(id)
, pModel(model)
, pSReacs()
{
    if (pModel == 0)
    {
        ostringstream os;
        os << "No model provided to Surfsys initializer function";
        throw steps::ArgErr(os.str());
    }
    pModel->_handleSurfsysAdd(this);
}

////////////////////////////////////////////////////////////////////////////////

Surfsys::~Surfsys(void)
{
    if (pModel == 0) return;
	_handleSelfDelete();
}

////////////////////////////////////////////////////////////////////////////////

void Surfsys::setID(string const & id)
{
    assert(pModel != 0);
    if (id == pID) return;
    // The following might raise an exception, e.g. if the new ID is not
    // valid or not unique. If this happens, we don't catch but simply let
    // it pass by into the Python layer.
    pModel->_handleSurfsysIDChange(pID, id);
    // This line will only be executed if the previous call didn't raise
    // an exception.
    pID = id;
}

////////////////////////////////////////////////////////////////////////////////

void Surfsys::_handleSelfDelete(void)
{
	std::vector<steps::model::SReac *> allsreacs = getAllSReacs();
	SReacPVecCI sreac_end = allsreacs.end();
	for(SReacPVecCI sreac = allsreacs.begin(); sreac != sreac_end; ++sreac)
	{
		delete(*sreac);
	}

	pModel->_handleSurfsysDel(this);
	pSReacs.clear();
	pModel = 0;
}

////////////////////////////////////////////////////////////////////////////////

SReac * Surfsys::getSReac(string const & id) const
{
    SReacPMapCI sreac = pSReacs.find(id);
    if (sreac == pSReacs.end())
    {
        ostringstream os;
        os << "Model does not contain surface "
        "reaction with name '" << id << "'";
        throw steps::ArgErr(os.str());
    }
    assert(sreac->second != 0);
    return sreac->second;
}

////////////////////////////////////////////////////////////////////////////////

void Surfsys::delSReac(string const & id)
{
    SReac * sreac = getSReac(id);
	delete(sreac);
}

////////////////////////////////////////////////////////////////////////////////

std::vector<SReac *> Surfsys::getAllSReacs(void) const
{
	SReacPVec sreacs = SReacPVec();
	SReacPMapCI sr_end = pSReacs.end();
	for (SReacPMapCI sr = pSReacs.begin(); sr != sr_end; ++sr)
	{
		sreacs.push_back(sr->second);
	}
	return sreacs;
}

////////////////////////////////////////////////////////////////////////////////

std::vector<Spec *> Surfsys::getAllSpecs(void) const
{
	SpecPVec specs = SpecPVec();
	bool first_occ = true;

	SReacPVec sreacs = getAllSReacs();
	SReacPVecCI sreac_end = sreacs.end();
	for (SReacPVecCI sreac = sreacs.begin(); sreac != sreac_end; ++sreac)
	{
		SpecPVec sr_specs = (*sreac)->getAllSpecs();
		SpecPVecCI sr_spec_end = sr_specs.end();
		for (SpecPVecCI sr_spec = sr_specs.begin();
		     sr_spec != sr_spec_end; ++sr_spec)
		{
			first_occ = true;
			SpecPVecCI allspecs_end = specs.end();
			for (SpecPVecCI allspecs = specs.begin();
				allspecs != allspecs_end; ++allspecs)
			{
				if ((*sr_spec) == (*allspecs))
				{
					first_occ = false;
					break;
				}
			}
			if (first_occ == true) specs.push_back(*sr_spec);
		}
	}

	return specs;
}

////////////////////////////////////////////////////////////////////////////////

void Surfsys::_checkSReacID(string const & id) const
{
    checkID(id);
    if (pSReacs.find(id) != pSReacs.end())
    {
        ostringstream os;
        os << "'" << id << "' is already in use";
        throw steps::ArgErr(os.str());
    }
}

////////////////////////////////////////////////////////////////////////////////

void Surfsys::_handleSReacIDChange(string const & o, string const & n)
{
    SReacPMapCI sr_old = pSReacs.find(o);
    assert(sr_old != pSReacs.end());

    if(o==n) return;
    _checkSReacID(n);

    SReac * sr = sr_old->second;
    assert(sr != 0);
    pSReacs.erase(sr->getID());
    pSReacs.insert(SReacPMap::value_type(n,sr));
}

////////////////////////////////////////////////////////////////////////////////

void Surfsys::_handleSReacAdd(SReac * sreac)
{
    assert(sreac->getSurfsys() == this);
    _checkSReacID(sreac->getID());
    pSReacs.insert(SReacPMap::value_type(sreac->getID(), sreac));
}

////////////////////////////////////////////////////////////////////////////////

void Surfsys::_handleSReacDel(SReac * sreac)
{
	assert (sreac->getSurfsys() == this);
    pSReacs.erase(sreac->getID());
}

////////////////////////////////////////////////////////////////////////////////

void Surfsys::_handleSpecDelete(Spec * spec)
{
	SReacPMapCI sreac_end = pSReacs.end();
	std::vector<std::string> sreacs_del = std::vector<std::string>();
    for (SReacPMapCI sreac = pSReacs.begin(); sreac != sreac_end; ++sreac)
    {
		SpecPVec specs = (sreac->second->getAllSpecs());
		SpecPVecCI sr_spec_end = specs.end();
		for (SpecPVecCI sr_spec = specs.begin();
			 sr_spec != sr_spec_end; ++sr_spec)
		{
			if ((*sr_spec) == spec)
			{
				sreacs_del.push_back(sreac->second->getID());
				break;
			}
		}
	}

	std::vector<std::string>::const_iterator sr_del_end = sreacs_del.end();
	for (std::vector<std::string>::const_iterator sr_del = sreacs_del.begin();
		 sr_del != sr_del_end; ++sr_del)
	{
		delSReac(*sr_del);
	}
}

////////////////////////////////////////////////////////////////////////////////

SReac * Surfsys::_getSReac(uint lidx) const
{
	assert (lidx < pSReacs.size());
	std::map<std::string, SReac *>::const_iterator sr_it = pSReacs.begin();
	for (uint i=0; i< lidx; ++i) ++sr_it;
	return sr_it->second;
}

////////////////////////////////////////////////////////////////////////////////

// END

File: test/unit/py/API_1/model_test/surfsys.py
####################################################################################
#
#    STEPS - STochastic Engine for Pathway Simulation
#    Copyright (C) 2007-2021 Okinawa Institute of Science and Technology, Japan.
#    Copyright (C) 2003-2006 University of Antwerp, Belgium.
#    
#    See the file AUTHORS for details.
#    This file is part of STEPS.
#    
#    STEPS is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License version 2,
#    as published by the Free Software Foundation.
#    
#    STEPS is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#    GNU General Public License for more details.
#    
#    You should have received a copy of the GNU General Public License
#    along with this program. If not, see <http://www.gnu.org/licenses/>.
#
#################################################################################   
###

""" Unit tests for steps.model.Surfsys class."""

import unittest
import random
from steps.model import *

class SurfsysCreationTestCase(unittest.TestCase):
    """ Tests for Surfsys construction. """
    def setUp(self):
        self.model = Model()
    
    def tearDown(self):
        self.model = None
    
    def testCreateSurfsys(self):
        ssys = Surfsys("ssys", self.model)
        self.assertIsInstance(ssys, Surfsys)


class SurfsysPropertyTestCase(unittest.TestCase):
    """ Property tests for Surfsys. """
    def setUp(self):
        self.model = Model()
        self.ssys = Surfsys("ssys", self.model)
    
    def tearDown(self):
        self.model = None
        self.ssys = None
    
    def testSetGetID(self):
        self.assertEqual(self.ssys.getID(), "ssys")
        self.ssys.setID("new_id")
        self.assertEqual(self.ssys.getID(), "new_id")

    def testGetModel(self):
        self.assertEqual(self.ssys.getModel().this, self.model.this)

class SurfsysSReacTestCase(unittest.TestCase):
    """ SReac tests for Surfsys. """
    def setUp(self):
        self.model = Model()
        self.ssys = Surfsys("ssys", self.model)
        self.spec_a = Spec("A", self.model)
        self.spec_b = Spec("B", self.model)
        self.spec_c = Spec("C", self.model)
        self.sreac1 = SReac("sreac1", self.ssys, ilhs = [self.spec_a], slhs = [self.spec_b], srhs = [self.spec_c], kcst = 1e-6)
        self.sreac2 = SReac("sreac2", self.ssys, ilhs = [self.spec_a], slhs = [self.spec_b], srhs = [self.spec_c], orhs = [self.spec_a], kcst = 1e-6)
    
    def tearDown(self):
        self.model = None
        self.ssys = None
        self.spec_a = None
        self.spec_b = None
        self.spec_c = None
        self.sreac1 = None
        self.sreac2 = None
    
    def testGetSReac(self):
        self.assertEqual(self.ssys.getSReac("sreac1").this, self.sreac1.this)
    
    def testDelSReac(self):
        self.ssys.delSReac("sreac1")
        self.assertRaises(Exception, self.ssys.getSReac, "sreac1")

    def testGetAllSReacs(self):
        stored_sreacs = [sreac.this for sreac in self.ssys.getAllSReacs()]
        expected_sreacs = [self.sreac1.this, self.sreac2.this]
        self.assertCountEqual(stored_sreacs, expected_sreacs)

class SurfsysDiffTestCase(unittest.TestCase):
    """ Diff tests for Surfsys. """
    def setUp(self):
        self.model = Model()
        self.ssys = Surfsys("ssys", self.model)
        self.spec_a = Spec("A", self.model)
        self.spec_b = Spec("B", self.model)
        self.diff1 = Diff("diff1", self.ssys, self.spec_a, 1e-6)
        self.diff2 = Diff("diff2", self.ssys, self.spec_b, 1e-6)
    
    def tearDown(self):
        self.model = None
        self.ssys = None
        self.spec_a = None
        self.spec_b = None
        self.diff1 = None
        self.diff2 = None
    
    def testGetDiff(self):
        self.assertEqual(self.ssys.getDiff("diff1").this, self.diff1.this)
    
    def testDelDiff(self):
        self.ssys.delDiff("diff1")
        self.assertRaises(Exception, self.ssys.getDiff, "diff1")
    
    def testGetAllDiffs(self):
        stored_diffs = [diff.this for diff in self.ssys.getAllDiffs()]
        expected_diffs = [self.diff1.this, self.diff2.this]
        self.assertCountEqual(stored_diffs, expected_diffs)

def suite():
    all_tests = []
    all_tests.append(unittest.makeSuite(SurfsysCreationTestCase, "test"))
    all_tests.append(unittest.makeSuite(SurfsysPropertyTestCase, "test"))
    all_tests.append(unittest.makeSuite(SurfsysSReacTestCase, "test"))
    all_tests.append(unittest.makeSuite(SurfsysDiffTestCase, "test"))
    return unittest.TestSuite(all_tests)

if __name__ == "__main__":
    unittest.TextTestRunner(verbosity=2).run(suite())
